pkg/sources/adapter/awssnssource/probe/probe.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package probe contains facilities for asserting the readiness of a
+// multi-tenant receive adapter.
+package probe
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/triggermesh/triggermesh/pkg/apis/sources/v1alpha1"
+	listersv1alpha1 "github.com/triggermesh/triggermesh/pkg/client/generated/listers/sources/v1alpha1"
+	"github.com/triggermesh/triggermesh/pkg/sources/adapter/common/router"
+)
+
+// ReadinessChecker can assert the readiness of a component.
+type ReadinessChecker interface {
+	IsReady() (bool, error)
+}
+
+// AdapterReadyChecker asserts the readiness of a receive adapter.
+type AdapterReadyChecker struct {
+	srcLister listersv1alpha1.AWSSNSSourceLister
+
+	srcRouter *router.Router
+
+	sync.RWMutex
+	isReady bool
+}
+
+var _ ReadinessChecker = (*AdapterReadyChecker)(nil)
+
+// NewAdapterReadyChecker returns an AdapterReadyChecker initialized with the
+// given Lister and Router.
+func NewAdapterReadyChecker(ls listersv1alpha1.AWSSNSSourceLister, r *router.Router) *AdapterReadyChecker {
+	return &AdapterReadyChecker{
+		srcLister: ls,
+		srcRouter: r,
+	}
+}
+
+// IsReady implements ReadinessChecker.
+// It checks whether the adapter has registered a handler for all observed sources.
+//
+// NOTE(antoineco): we might want to revisit this in the future, because the
+// following implementation details are currently leaking into this
+// ReadinessChecker implementation:
+//
+// - The fact that the adapter exposes a health endpoint via the HTTP router.
+// - The exact URL path of this health endpoint.
+// - The fact that a handler isn't registered for sources which don't have a valid sink.
+//
+// We consider this tight coupling acceptable for the time being, because this
+// is a sub-package of the main adapter package for the SNS source
+// specifically.
+func (c *AdapterReadyChecker) IsReady() (bool, error) {
+	// readiness already observed at an earlier point in time,
+	// short-circuit the check
+	if c.readLockedIsReady() {
+		return true, nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	// double-checked lock to ensure we don't write the value of "isReady"
+	// twice if multiple goroutines called IsReady() simultaneously.
+	if c.isReady {
+		return true, nil
+	}
```

Comment:
Just in case that part isn't clear, the double check on `c.isReady` guarantees that we avoid continuing unnecessarily in case the following scenario occurs:

```golang
/* Two concurrent calls to IsReady() cause two goroutines to land here simultaneously: */

	// <-------- goroutine1
	// <-------- goroutine2
	c.Lock()
	defer c.Unlock()
```

```golang
/* The mutex lets *one* routine through while the other waits until we call Unlock(): */

	// <-------- goroutine2
	c.Lock()
	defer c.Unlock()
	// <-------- goroutine1
```

```golang
/* The function called by goroutine1 returned, isReady is now true, but goroutine2 is still
    inside IsReady() and gets unblocked: */

	c.Lock()
	defer c.Unlock()

	// <-------- goroutine2

	// we save goroutine2 the effort of doing again what goroutine1 just did
	if c.isReady {
		return true, nil
	}
```